### ISSUE TITLE
Option for arithmetics to ignore nans created by alignment

### ISSUE THREAD

**Issue by crusaderky** [pydata/xarray#1625]
Can anybody tell me if there is anybody who benefits from this behaviour? I can't think of any good use cases.

```
wallet = xarray.DataArray([50, 70], dims=['currency'], coords={'currency': ['EUR', 'USD']})
restaurant_bill = xarray.DataArray([30], dims=['currency'], coords={'currency': ['USD']})
with xarray.set_options(arithmetic_join="outer"):
    print(wallet - restaurant_bill)

<xarray.DataArray (currency: 2)>
array([ nan,  40.])
Coordinates:
  * currency  (currency) object 'EUR' 'USD'
```

While it is fairly clear why it can be desirable to have ``nan + not nan = nan`` as a default in arithmetic when the nan is already present in one of the input arrays, when the nan is introduced as part of an automatic align things become much less intuitive.

Proposal:
- add a parameter to ``xarray.align``, ``fillvalue=numpy.nan``, which determines what will appear in the newly created array elements
- change \_\_add\_\_, \_\_sub\_\_ etc. to invoke ``xarray.align(fillvalue=0)``
- change \_\_mul\_\_, \_\_truediv\_\_ etc. to invoke ``xarray.align(fillvalue=1)``

In theory the setting could be left as an opt-in as ``set_options(arithmetic_align_fillvalue='neutral')``, yet I wonder who would actually want the current behaviour?

**Comment by crusaderky** [comment]
Related pure-numpy thread
https://stackoverflow.com/questions/42209838/treat-nan-as-zero-in-numpy-array-summation-except-for-nan-in-all-arrays

**Comment by shoyer** [comment]
This behavior is consistent with the default behavior on pandas, which always does an outer join for arithmetic:
```
In [6]: wallet.to_series() - restaurant_bill.to_series()
currency
EUR     NaN
USD    40.0
dtype: float64
```
So I don't think we want to change this in an inconsistent way in xarray.

I do agree that we should have support for an explicit fill value in alignment/reindexing and arithmetic. For consistency with pandas (and elsewhere in xarray), let's call it `fill_value`.

> - change `__add__`,` __sub__` etc. to invoke `xarray.align(fillvalue=0)`
> - change `__mul__`, `__truediv__` etc. to invoke `xarray.align(fillvalue=1)`

I can see the logic in using an identity value instead of NaN as a default in arithmetic. One peril of this approach is that it isn't always evident what the right identity is. In fact, according to NumPy:
```
In [16]: import numpy as np

In [17]: print(np.add.identity)
0

In [18]: print(np.multiply.identity)
1

In [19]: print(np.subtract.identity)
None

In [20]: print(np.divide.identity)
None
```

Let me give a couple other examples of why I don't think we should use an identity of some sort as the default:
- Suppose we are comparing a model to observations. `model - observations` gives the residuals -- unless the variables are mis-aligned, in which case it would give a residual equal to one of the variables.
- Suppose we want to do a weighted average. `(observations * weights).sum()` gives a sensible answer -- but not if a missing weight defaults to 1 (all the weights put together are supposed to sum to 1!).

**Comment by crusaderky** [comment]
Updated title. I'm happy with the explicit approach.
New design prototype:
```
wallet, restaurant_bill = xarray.align(wallet, restaurant_bill, join='outer', fill_value=0)
print(wallet - restaurant_bill)

<xarray.DataArray (currency: 2)>
array([50, 40])
Coordinates:
  * currency  (currency) object 'EUR' 'USD'
```

**Comment by crusaderky** [comment]
Closed by #2876
